src/ripple/rpc/handlers/Catalogue.cpp

```diff
@@ -0,0 +1,1139 @@
+
+            ledger->setLedgerInfo(info);
+
+            // Deserialize the complete state map from leaf nodes
+            if (!ledger->stateMap().deserializeFromStream(*decompStream))
+            {
+                JLOG(context.j.error())
+                    << "Failed to deserialize base ledger state";
+                return rpcError(
+                    rpcINTERNAL, "Failed to load base ledger state");
+            }
+        }
+        else
+        {
+            // Delta ledger - start with a copy of the previous ledger
+            if (!prevLedger)
+            {
+                JLOG(context.j.error()) << "Missing previous ledger for delta";
+                return rpcError(rpcINTERNAL, "Missing previous ledger");
+            }
+
+            auto snapshot = prevLedger->stateMap().snapShot(true);
+
+            ledger = std::make_shared<Ledger>(
+                info,
+                context.app.config(),
+                context.app.getNodeFamily(),
+                *snapshot);
+
+            // Apply delta (only leaf-node changes)
+            if (!ledger->stateMap().deserializeFromStream(*decompStream))
+            {
+                JLOG(context.j.error())
+                    << "Failed to apply delta to ledger " << info.seq;
+                return rpcError(rpcINTERNAL, "Failed to apply ledger delta");
+            }
+        }
+
+        // pull in the tx map
+        if (!ledger->txMap().deserializeFromStream(*decompStream))
+        {
+            JLOG(context.j.error())
+                << "Failed to apply delta to ledger " << info.seq;
+            return rpcError(rpcINTERNAL, "Failed to apply ledger delta");
+        }
+
+        // Finalize the ledger
+        ledger->stateMap().flushDirty(hotACCOUNT_NODE);
+        ledger->txMap().flushDirty(hotTRANSACTION_NODE);
+
+        ledger->setAccepted(
+            info.closeTime,
+            info.closeTimeResolution,
+            info.closeFlags & sLCF_NoConsensusTime);
+
+        ledger->setValidated();
+        ledger->setCloseFlags(info.closeFlags);
+        ledger->setImmutable(true);
+
+        // Save in database
+        pendSaveValidated(context.app, ledger, false, false);
+
+        // Store in ledger master
+        context.app.getLedgerMaster().storeLedger(ledger, true);
+
+        if (info.seq == header.max_ledger &&
+            context.app.getLedgerMaster().getClosedLedger()->info().seq <
+                info.seq)
+        {
+            // Set as current ledger if this is the latest
+            context.app.getLedgerMaster().switchLCL(ledger);
+        }
+
+        context.app.getLedgerMaster().setLedgerRangePresent(
+            header.min_ledger, info.seq, true);
+
+        // Store the ledger
+        prevLedger = ledger;
+        ledgersLoaded++;
+    }
+
+    decompStream->reset();
+    infile.close();
+
+    // Update ledger range in ledger master
+    context.app.getLedgerMaster().setLedgerRangePresent(
```

Comment:
```C++
        context.app.getLedgerMaster().setLedgerRangePresent(
            header.min_ledger, info.seq, true);

        // Store the ledger
        prevLedger = ledger;
        ledgersLoaded++;
    }

    decompStream->reset();
    infile.close();

    // Update ledger range in ledger master
    context.app.getLedgerMaster().setLedgerRangePresent(
        header.min_ledger, header.max_ledger, true);
```

The last one seems redundant if it's already called as each ledger is loaded ? Actually, if you look at the loop, there's only one break rather than an early return, and in that case, it assumes success? 
And marks the full header's ledger range?